### ISSUE TITLE
Need help in understanding the logic

### ISSUE THREAD

**Issue by dhirajsuvarna** [tensorsense/Retrieval-Framework#5]
First of all thanks for this notebook, it helps in understanding how to build the nodes structure to preform good retrievals. 

Having said that I wanted to understand some reasoning behind some of the code. 

Following are my queries - 
1. What is the significance of `start_char_idx` and `end_char_idx` in the `TextNode`. Is it mandatory to set it for retrieval?
2. What is the reasoning behind creating Node Relationships? why is it important?

I dont know if this is the correct forum to ask these questions, if not, can you please point me to correct forum?

**Comment by anbuzin** [comment]
Hey @dhirajsuvarna!

1. These are not mandatory, at least not for this particular retriever setup. But they do enable all kinds of customizations by linking the node to its position in the original text. I decided to keep this information since it was already at hand, but for this setup it's redundant.
2. Same as the previous answer: redundant for this setup, but enables other retrievers should you need them.

**Comment by dhirajsuvarna** [comment]
@deepbuzin  - Thanks

**Comment by dhirajsuvarna** [comment]
Continuing on this, have you been able to perisit the nodes in any VectoStore and then load it from the persisted storage to rebuild the RecursiveRetriever?

**Comment by anbuzin** [comment]
@dhirajsuvarna yeah, it can be a little confusing. Vector stores only store embeddings, LlamaIndex handles saving the index, and you have to take care of persisting full nodes separately. All LlamaIndex nodes are Pydantic objects, so they are pretty easy to store and load:

```python
# save to disk

chroma_client = chromadb.PersistentClient(path=persist_path.as_posix())
chroma_collection = chroma_client.get_or_create_collection(collection_name)

vector_store = ChromaVectorStore(chroma_collection=chroma_collection)
storage_context = StorageContext.from_defaults(vector_store=vector_store)

index = VectorStoreIndex(
    stored_nodes, storage_context=storage_context, show_progress=True
)

full_nodes_path = persist_path.joinpath("full_nodes.json")

serializable = {k: v.json() for k, v in full_nodes_dict.items()}
with full_nodes_path.open("w") as f:
    json.dump(serializable, f)
```

```python
# load from disk
chroma_client = chromadb.PersistentClient(path=persist_path.as_posix())
chroma_collection = chroma_client.get_or_create_collection(collection_name)

vector_store = ChromaVectorStore(chroma_collection=chroma_collection)
index = VectorStoreIndex.from_vector_store(
    vector_store,
)

full_nodes_path = persist_path.joinpath("full_nodes.json")
with full_nodes_path.open("r") as f:
    serializable = json.load(f)

full_nodes_dict = {k: Document.parse_raw(v) for k, v in serializable.items()}
```

And then you would rebuild the retriever same way it was originally built.
